<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

# Working with a real world data-set using SQL and Python

Estimated time needed: **30** minutes

## Objectives

After completing this lab you will be able to:

*   Understand the dataset for Chicago Public School level performance
*   Store the dataset in SQLite database.
*   Retrieve metadata about tables and columns and query data from mixed case columns
*   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&download=true&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**:

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">link</a>.

Now review some of its contents.


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

##### The syntax for connecting to magic sql using sqllite is

  **%sql sqlite://DatabaseName**   
  
where DatabaseName will be your **.db** file 


In [6]:
import csv, sqlite3

con = sqlite3.connect("RealWorldData.db")
cur = con.cursor()

In [7]:
!pip install -q pandas==1.1.5

  error: subprocess-exited-with-error
  
  × Building wheel for pandas (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [1155 lines of output]
      <string>:19: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
      <string>:45: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
      running bdist_wheel
      running build
      running build_py
      creating build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/_typing.py -> build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/conftest.py -> build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/testing.py -> build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/_version.py -> build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/__init__.py -> build/lib.linux-x86_64-cpython-311/pandas
      copying pandas/_testing.py -> build/lib.linux-x86_64-cpython-311/panda

In [3]:
%load_ext sql

In [4]:
%sql sqlite:///RealWorldData.db

### Store the dataset in a Table

##### In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet.  
##### To analyze the data using SQL, it first needs to be stored in the database.

##### We will first read the csv files  from the given url  into pandas dataframes

##### Next we will be using the  df.to_sql() function to convert each csv file  to a table in sqlite  with the csv data loaded in it.


In [5]:
import pandas as pd

df = pd.read_csv('ChicagoPublicSchools.csv')
df.to_sql("ChicagoPublicSchools", con, if_exists = 'replace', index = False, method = "multi")

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/pandas/core/generic.py:2615: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  method=method,


Double-click **here** for the solution.
<!--

import pandas

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", con, if_exists='replace', index=False, method="multi")

-->


### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [6]:
# type in your query to retrieve list of all tables in the database
%sql SELECT * FROM ChicagoPublicSchools limit 10;

 * sqlite:///RealWorldData.db
Done.


[(610038, 'Abraham Lincoln Elementary School', 'ES', '615 W Kemper Pl', 'Chicago', 'IL', 60614, '(773) 534-5720', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610038.pdf', 'Fullerton Elementary Network', 'NORTH-NORTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 1', 'Yes', 'Very Strong', 99.0, 'Very Strong', '99', 'Strong', 74.0, 'Strong', 66.0, 'Weak', '65', 'Strong', '70', 'Strong', '56', 'Average', '47', '96.00%', 2.0, '96.40%', '95.80%', '80.1', '43.3', '89.6', '84.9', '60.7', '62.6', '81.9', '85.2', '52', '62.4', '66.3', '77.9', 69.7, 64.4, 0.2, 0.9, 'Yellow', 'Green', '67.1', '54.5', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 813, 33, 'NDA', 1171699.4579999999, 1915829.428, 41.92449696, -87.64452163, 7, 'LINCOLN PARK', 43, 18, '(41.92449696, -87.64452163)'),
 (610281, 'Adam Clayton Powell Paideia Community Academy Elementary School', 'ES', '7511 S South Shore Dr', 'Chicago', 'IL', 60649, '(773) 535-6650', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610281.pdf', 'Skyway Elementary Network', 'SOUTH SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 54.0, 'Strong', '66', 'Strong', 74.0, 'Very Strong', 84.0, 'Weak', '63', 'Strong', '76', 'Weak', '46', 'Average', '50', '95.60%', 15.7, '95.30%', '100.00%', '62.4', '51.7', '21.9', '15.1', '29', '42.8', '38.5', '27.4', '44.8', '42.7', '14.1', '34.4', 16.8, 16.5, 0.7, 1.4, 'Green', 'Green', '17.2', '27.3', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 521, 46, 'NDA', 1196129.985, 1856209.466, 41.76032435, -87.55673627, 43, 'SOUTH SHORE', 7, 4, '(41.76032435, -87.55673627)'),
 (610185, 'Adlai E Stevenson Elementary School', 'ES', '8010 S Kostner Ave', 'Chicago', 'IL', 60652, '(773) 535-2280', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610185.pdf', 'Midway Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Standard', 'Not on Probation', 'Level 2', 'No', 'Strong', 61.0, 'NDA', 'NDA', 'Average', 50.0, 'Weak', 36.0, 'Weak', 'NDA', 'NDA', 'NDA', 'Average', '47', 'Weak', '41', '95.70%', 2.3, '94.70%', '98.30%', '53.7', '26.6', '38.3', '34.7', '43.7', '57.3', '48.8', '39.2', '46.8', '44', '7.5', '21.9', 18.3, 15.5, -0.9, -1.0, 'Red', 'Red', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 1324, 44, 'NDA', 1148427.165, 1851012.215, 41.74711093, -87.73170248, 70, 'ASHBURN', 13, 8, '(41.74711093, -87.73170248)'),
 (609993, 'Agustin Lara Elementary Academy', 'ES', '4619 S Wolcott Ave', 'Chicago', 'IL', 60609, '(773) 535-4389', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_609993.pdf', 'Pershing Elementary Network', 'SOUTHWEST SIDE COLLABORATIVE', 'No', 'Track_E', 'Not on Probation', 'Level 1', 'No', 'Average', 56.0, 'Average', '44', 'Average', 45.0, 'Weak', 37.0, 'Weak', '65', 'Average', '48', 'Average', '53', 'Strong', '58', '95.50%', 10.4, '95.80%', '100.00%', '76.9', 'NDA', '26', '24.7', '61.8', '49.7', '39.2', '27.2', '69.7', '60.6', '9.1', '18.2', 11.1, 9.6, 0.9, 2.4, 'Green', 'Green', '42.9', '25', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 556, 42, 'NDA', 1164504.29, 1873959.199, 41.8097569, -87.6721446, 61, 'NEW CITY', 20, 9, '(41.8097569, -87.6721446)'),
 (610513, 'Air Force Academy High School', 'HS', '3630 S Wells St', 'Chicago', 'IL', 60609, '(773) 535-1590', 'http://schoolreports.cps.edu/SchoolProgressReport_Eng/Spring2011Eng_610513.pdf', 'Southwest Side High School Network', 'SOUTHWEST SIDE COLLABORATIVE', 'NDA', 'Standard', 'Not on Probation', 'Not Enough Data', 'Yes', 'Average', 49.0, 'Strong', '60', 'Strong', 60.0, 'Average', 55.0, 'Weak', '45', 'Average', '54', 'Average', '53', 'Average', '49', '93.30%', 15.6, '96.90%', '100.00%', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', 'NDA', None, None, None, None, 'NDA', 'NDA', 'NDA', 'NDA', '14.6', '14.8', 'NDA', '16', '1.4', 

Double-click **here** for a hint

<!--
In sqlite the system catalog table called sqlite_master contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution using SQLite:

%sql SELECT name FROM sqlite_master WHERE type='table'

-->



### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [7]:
# type in your query to retrieve the number of columns in the SCHOOLS table
%sql select count(*) from ChicagoPublicSchools;

 * sqlite:///RealWorldData.db
Done.


[(566,)]

Double-click **here** for the solution.

<!-- Solution SQLite:
%sql SELECT count(name) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->




Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.


In [8]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length
%sql PRAGMA TABLE_INFO([ChicagoPublicSchools])

 * sqlite:///RealWorldData.db
Done.


[(0, 'School_ID', 'INTEGER', 0, None, 0),
 (1, 'NAME_OF_SCHOOL', 'TEXT', 0, None, 0),
 (2, 'Elementary, Middle, or High School', 'TEXT', 0, None, 0),
 (3, 'Street_Address', 'TEXT', 0, None, 0),
 (4, 'City', 'TEXT', 0, None, 0),
 (5, 'State', 'TEXT', 0, None, 0),
 (6, 'ZIP_Code', 'INTEGER', 0, None, 0),
 (7, 'Phone_Number', 'TEXT', 0, None, 0),
 (8, 'Link', 'TEXT', 0, None, 0),
 (9, 'Network_Manager', 'TEXT', 0, None, 0),
 (10, 'Collaborative_Name', 'TEXT', 0, None, 0),
 (11, 'Adequate_Yearly_Progress_Made_', 'TEXT', 0, None, 0),
 (12, 'Track_Schedule', 'TEXT', 0, None, 0),
 (13, 'CPS_Performance_Policy_Status', 'TEXT', 0, None, 0),
 (14, 'CPS_Performance_Policy_Level', 'TEXT', 0, None, 0),
 (15, 'HEALTHY_SCHOOL_CERTIFIED', 'TEXT', 0, None, 0),
 (16, 'Safety_Icon', 'TEXT', 0, None, 0),
 (17, 'SAFETY_SCORE', 'REAL', 0, None, 0),
 (18, 'Family_Involvement_Icon', 'TEXT', 0, None, 0),
 (19, 'Family_Involvement_Score', 'TEXT', 0, None, 0),
 (20, 'Environment_Icon', 'TEXT', 0, None, 0),
 (21, 'Environment_Score', 'REAL', 0, None, 0),
 (22, 'Instruction_Icon', 'TEXT', 0, None, 0),
 (23, 'Instruction_Score', 'REAL', 0, None, 0),
 (24, 'Leaders_Icon', 'TEXT', 0, None, 0),
 (25, 'Leaders_Score', 'TEXT', 0, None, 0),
 (26, 'Teachers_Icon', 'TEXT', 0, None, 0),
 (27, 'Teachers_Score', 'TEXT', 0, None, 0),
 (28, 'Parent_Engagement_Icon', 'TEXT', 0, None, 0),
 (29, 'Parent_Engagement_Score', 'TEXT', 0, None, 0),
 (30, 'Parent_Environment_Icon', 'TEXT', 0, None, 0),
 (31, 'Parent_Environment_Score', 'TEXT', 0, None, 0),
 (32, 'AVERAGE_STUDENT_ATTENDANCE', 'TEXT', 0, None, 0),
 (33, 'Rate_of_Misconducts__per_100_students_', 'REAL', 0, None, 0),
 (34, 'Average_Teacher_Attendance', 'TEXT', 0, None, 0),
 (35, 'Individualized_Education_Program_Compliance_Rate', 'TEXT', 0, None, 0),
 (36, 'Pk_2_Literacy__', 'TEXT', 0, None, 0),
 (37, 'Pk_2_Math__', 'TEXT', 0, None, 0),
 (38, 'Gr3_5_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (39, 'Gr3_5_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (40, 'Gr3_5_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (41, 'Gr3_5_Keep_Pace_Math__', 'TEXT', 0, None, 0),
 (42, 'Gr6_8_Grade_Level_Math__', 'TEXT', 0, None, 0),
 (43, 'Gr6_8_Grade_Level_Read__', 'TEXT', 0, None, 0),
 (44, 'Gr6_8_Keep_Pace_Math_', 'TEXT', 0, None, 0),
 (45, 'Gr6_8_Keep_Pace_Read__', 'TEXT', 0, None, 0),
 (46, 'Gr_8_Explore_Math__', 'TEXT', 0, None, 0),
 (47, 'Gr_8_Explore_Read__', 'TEXT', 0, None, 0),
 (48, 'ISAT_Exceeding_Math__', 'REAL', 0, None, 0),
 (49, 'ISAT_Exceeding_Reading__', 'REAL', 0, None, 0),
 (50, 'ISAT_Value_Add_Math', 'REAL', 0, None, 0),
 (51, 'ISAT_Value_Add_Read', 'REAL', 0, None, 0),
 (52, 'ISAT_Value_Add_Color_Math', 'TEXT', 0, None, 0),
 (53, 'ISAT_Value_Add_Color_Read', 'TEXT', 0, None, 0),
 (54, 'Students_Taking__Algebra__', 'TEXT', 0, None, 0),
 (55, 'Students_Passing__Algebra__', 'TEXT', 0, None, 0),
 (56, '9th Grade EXPLORE (2009)', 'TEXT', 0, None, 0),
 (57, '9th Grade EXPLORE (2010)', 'TEXT', 0, None, 0),
 (58, '10th Grade PLAN (2009)', 'TEXT', 0, None, 0),
 (59, '10th Grade PLAN (2010)', 'TEXT', 0, None, 0),
 (60, 'Net_Change_EXPLORE_and_PLAN', 'TEXT', 0, None, 0),
 (61, '11th Grade Average ACT (2011)', 'TEXT', 0, None, 0),
 (62, 'Net_Change_PLAN_and_ACT', 'TEXT', 0, None, 0),
 (63, 'College_Eligibility__', 'TEXT', 0, None, 0),
 (64, 'Graduation_Rate__', 'TEXT', 0, None, 0),
 (65, 'College_Enrollment_Rate__', 'TEXT', 0, None, 0),
 (66, 'COLLEGE_ENROLLMENT', 'INTEGER', 0, None, 0),
 (67, 'General_Services_Route', 'INTEGER', 0, None, 0),
 (68, 'Freshman_on_Track_Rate__', 'TEXT', 0, None, 0),
 (69, 'X_COORDINATE', 'REAL', 0, None, 0),
 (70, 'Y_COORDINATE', 'REAL', 0, None, 0),
 (71, 'Latitude', 'REAL', 0, None, 0),
 (72, 'Longitude', 'REAL', 0, None, 0),
 (73, 'COMMUNITY_AREA_NUMBER', 'INTEGER', 0, None, 0),
 (74, 'COMMUNITY_AREA_NAME', 'TEXT', 0, None, 0),
 (75, 'Ward', 'INTEGER', 0, None, 0),
 (76, 'Police_District', 'INTEGER', 0, None, 0),
 (77, 'Location', 'TEXT', 0, None, 0)]

Double-click **here** for the solution.

<!-- Solution 

%sql SELECT name,type,length(type) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->




### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "\_"?


## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [16]:
%sql select count(*) from ChicagoPublicSchools where "Elementary, Middle, or High School" = 'MS'

 * sqlite:///RealWorldData.db
Done.


[(11,)]

Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from CHICAGO_PUBLIC_SCHOOLS_DATA where "Elementary, Middle, or High School"='ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [9]:

%sql select max(SAFETY_SCORE) from ChicagoPublicSchools

 * sqlite:///RealWorldData.db
Done.


[(99.0,)]

Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from CHICAGO_PUBLIC_SCHOOLS_DATA

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [19]:
%sql select NAME_OF_SCHOOL from ChicagoPublicSchools where SAFETY_SCORE = 99.0


 * sqlite:///RealWorldData.db
Done.


[('Abraham Lincoln Elementary School',),
 ('Alexander Graham Bell Elementary School',),
 ('Annie Keller Elementary Gifted Magnet School',),
 ('Augustus H Burley Elementary School',),
 ('Edgar Allan Poe Elementary Classical School',),
 ('Edgebrook Elementary School',),
 ('Ellen Mitchell Elementary School',),
 ('James E McDade Elementary Classical School',),
 ('James G Blaine Elementary School',),
 ('LaSalle Elementary Language Academy',),
 ('Mary E Courtenay Elementary Language Arts Center',),
 ('Northside College Preparatory High School',),
 ('Northside Learning Center High School',),
 ('Norwood Park Elementary School',),
 ('Oriole Park Elementary School',),
 ('Sauganash Elementary School',),
 ('Stephen Decatur Classical Elementary School',),
 ('Talman Elementary School',),
 ('Wildwood Elementary School',)]

Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where \
  Safety_Score= (select MAX(Safety_Score) from CHICAGO_PUBLIC_SCHOOLS_DATA)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [21]:
%sql select NAME_OF_SCHOOL, AVERAGE_STUDENT_ATTENDANCE from ChicagoPublicSchools order by AVERAGE_STUDENT_ATTENDANCE desc limit 10

 * sqlite:///RealWorldData.db
Done.


[('John Charles Haines Elementary School', '98.40%'),
 ('James Ward Elementary School', '97.80%'),
 ('Edgar Allan Poe Elementary Classical School', '97.60%'),
 ('Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 ('Rachel Carson Elementary School', '97.60%'),
 ('Annie Keller Elementary Gifted Magnet School', '97.50%'),
 ('Andrew Jackson Elementary Language Academy', '97.40%'),
 ('Lenart Elementary Regional Gifted Center', '97.40%'),
 ('Disney II Magnet School', '97.30%'),
 ('John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from CHICAGO_PUBLIC_SCHOOLS_DATA \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [22]:
%sql select NAME_OF_SCHOOL, AVERAGE_STUDENT_ATTENDANCE from ChicagoPublicSchools order by AVERAGE_STUDENT_ATTENDANCE limit 5

 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [24]:
%sql select NAME_OF_SCHOOL, REPLACE(AVERAGE_STUDENT_ATTENDANCE,'%','') from ChicagoPublicSchools order by AVERAGE_STUDENT_ATTENDANCE limit 5


 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Dyett High School', '62.50'),
 ('Wendell Phillips Academy High School', '63.00')]

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''

-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [11]:
%sql select NAME_OF_SCHOOL, AVERAGE_STUDENT_ATTENDANCE from ChicagoPublicSchools where AVERAGE_STUDENT_ATTENDANCE < '70%'

 * sqlite:///RealWorldData.db
Done.


[('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Chicago Vocational Career Academy High School', '68.80%'),
 ('Dyett High School', '62.50%'),
 ('Manley Career Academy High School', '66.80%'),
 ('Orr Academy High School', '66.30%'),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Roberto Clemente Community Academy High School', '69.60%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)

-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [17]:
%sql select COMMUNITY_AREA_NAME, COLLEGE_ENROLLMENT, count(*) from ChicagoPublicSchools group by COMMUNITY_AREA_NAME

 * sqlite:///RealWorldData.db
Done.


[('ALBANY PARK', 266, 8),
 ('ARCHER HEIGHTS', 3320, 2),
 ('ARMOUR SQUARE', 302, 3),
 ('ASHBURN', 1324, 8),
 ('AUBURN GRESHAM', 602, 10),
 ('AUSTIN', 316, 23),
 ('AVALON PARK', 398, 3),
 ('AVONDALE', 644, 4),
 ('BELMONT CRAGIN', 634, 12),
 ('BEVERLY', 274, 4),
 ('BRIDGEPORT', 600, 5),
 ('BRIGHTON PARK', 691, 8),
 ('BURNSIDE', 549, 1),
 ('CALUMET HEIGHTS', 264, 6),
 ('CHATHAM', 652, 9),
 ('CHICAGO LAWN', 531, 7),
 ('CLEARING', 137, 4),
 ('DOUGLAS', 483, 11),
 ('DUNNING', 784, 6),
 ('EAST GARFIELD PARK', 551, 13),
 ('EAST SIDE', 632, 5),
 ('EDGEWATER', 709, 6),
 ('EDISON PARK', 650, 2),
 ('ENGLEWOOD', 539, 17),
 ('FOREST GLEN', 468, 3),
 ('FULLER PARK', 224, 2),
 ('GAGE PARK', 1021, 10),
 ('GARFIELD RIDGE', 1589, 5),
 ('GRAND BOULEVARD', 436, 8),
 ('GREATER GRAND CROSSING', 310, 10),
 ('HEGEWISCH', 633, 2),
 ('HERMOSA', 800, 4),
 ('HUMBOLDT PARK', 801, 13),
 ('HYDE PARK', 340, 4),
 ('IRVING PARK', 2366, 9),
 ('JEFFERSON PARK', 574, 2),
 ('KENWOOD', 550, 7),
 ('LAKE VIEW', 307, 11),
 ('LINCOLN PARK', 813, 7),
 ('LINCOLN SQUARE', 451, 5),
 ('LOGAN SQUARE', 586, 11),
 ('LOOP', 871, 1),
 ('LOWER WEST SIDE', 1683, 11),
 ('MCKINLEY PARK', 321, 4),
 ('MONTCLARE', 1317, 1),
 ('MORGAN PARK', 204, 5),
 ('MOUNT GREENWOOD', 245, 4),
 ('NEAR NORTH SIDE', 314, 7),
 ('NEAR SOUTH SIDE', 446, 3),
 ('NEAR WEST SIDE', 556, 16),
 ('NEW CITY', 556, 13),
 ('NORTH CENTER', 4368, 7),
 ('NORTH LAWNDALE', 171, 16),
 ('NORTH PARK', 1636, 5),
 ('NORWOOD PARK', 160, 8),
 ('OAKLAND', 140, 1),
 ('OHARE', 786, 1),
 ('PORTAGE PARK', 1716, 7),
 ('PULLMAN', 192, 5),
 ('RIVERDALE', 532, 4),
 ('ROGERS PARK', 375, 6),
 ('ROSELAND', 525, 13),
 ('SOUTH CHICAGO', 905, 7),
 ('SOUTH DEERING', 286, 4),
 ('SOUTH LAWNDALE', 446, 22),
 ('SOUTH SHORE', 521, 9),
 ('UPTOWN', 280, 7),
 ('WASHINGTON HEIGHTS', 246, 9),
 ('WASHINGTON PARK', 350, 5),
 ('WEST ELSDON', 1560, 3),
 ('WEST ENGLEWOOD', 339, 13),
 ('WEST GARFIELD PARK', 430, 8),
 ('WEST LAWN', 830, 4),
 ('WEST PULLMAN', 138, 10),
 ('WEST RIDGE', 752, 9),
 ('WEST TOWN', 620, 20),
 ('WOODLAWN', 730, 8)]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [18]:
%sql select COMMUNITY_AREA_NAME, COLLEGE_ENROLLMENT, count(*) from ChicagoPublicSchools group by COMMUNITY_AREA_NAME order by COLLEGE_ENROLLMENT limit 5

 * sqlite:///RealWorldData.db
Done.


[('CLEARING', 137, 4),
 ('WEST PULLMAN', 138, 10),
 ('OAKLAND', 140, 1),
 ('NORWOOD PARK', 160, 8),
 ('NORTH LAWNDALE', 171, 16)]

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   LIMIT 5 

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [32]:
%sql select NAME_OF_SCHOOL, SAFETY_SCORE from ChicagoPublicSchools where SAFETY_SCORE != 'Null' order by SAFETY_SCORE limit 5

 * sqlite:///RealWorldData.db
Done.


[('Edmond Burke Elementary School', 1.0),
 ("Luke O'Toole Elementary School", 5.0),
 ('George W Tilton Elementary School', 6.0),
 ('Foster Park Elementary School', 11.0),
 ('Emil G Hirsch Metropolitan High School', 13.0)]

Double-click **here** for the solution.

<!-- Solution SQLite:

%sql SELECT name_of_school, safety_score \
FROM CHICAGO_PUBLIC_SCHOOLS_DATA  where safety_score !='None' \
ORDER BY safety_score \
LIMIT 5

-->




### Problem 11

##### Get the hardship index for the community area of the school which has College Enrollment of 4368


In [ ]:
%sql select hardship index from 


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist. If the CHICAGO_SOCIOECONOMIC_DATA table does not exists, please execute the following code beforehand.

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")


%%sql 
select hardship_index from CENSUS_DATA CD, CHICAGO_PUBLIC_SCHOOLS_DATA CPS 
where CD.community_area_number = CPS.community_area_number 
and college_enrollment = 4368


-->


### Problem 12

##### Get the hardship index for the community area which has the highest value for College Enrollment


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist. If the CHICAGO_SOCIOECONOMIC_DATA table does not exists, please execute the following code beforehand.

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")

%sql select community_area_number, community_area_name, hardship_index from CENSUS_DATA \
   where community_area_number in \
   ( select community_area_number from CHICAGO_PUBLIC_SCHOOLS_DATA order by college_enrollment desc limit 1 )
-->


## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Rav Ahuja</a>
<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                        |
| ----------------- | ------- | ----------------- | ----------------------------------------- |
| 2022-03-04        | 2.2     | Lakshmi Holla     | Made changes in markdown cells |
| 2020-11-27        | 2.1     | Sannareddy Ramesh | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya           | Moved lab to course repo in GitLab        |
-->
<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
